polyply/src/nonbond_matrix.py

```diff
@@ -0,0 +1,247 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import itertools
+import multiprocessing
+import numpy as np
+import scipy.spatial
+from numba import jit
+from .topology import lorentz_berthelot_rule
+
+
+def _lennard_jones_force(dist, point, ref, params):
+    """
+    Compute the Lennard-Jones force between two particles
+    given their interaction parameters, the distance, and
+    distance vector.
+
+    Parameters:
+    -----------
+    dist: float
+        the distance between particles
+    vect: np.ndarray(3)
+        the distance vector
+    params: tuple
+        tuple of sigma and epsilon parameters
+
+    Returns:
+    --------
+    numpy.ndarray(3)
+        the force vector
+    """
+    sig, eps = params
+    # doing this here with numba is faster than separate
+    vect = point - ref
+    force = 24 * eps / dist * ((2 * (sig/dist)**12.0) - (sig/dist)**6) * vect/dist
+    return force
+lennard_jones_force = jit(_lennard_jones_force, nopython=True, cache=True, fastmath=True)
+
+
+POTENTIAL_FUNC = {"LJ": lennard_jones_force}
+
+
+def _n_particles(molecules):
+    """
+    Count the number of meta_molecule nodes
+    in the topology.
+    """
+    n_atoms = 0
+    for molecule in molecules:
+        n_atoms += len(molecule.nodes)
+    return n_atoms
+
+class NonBondMatrix():
+    """
+    Object for handeling nonbonded interactions
+    of the topology. It stores the positions using
+    infinity for undefined positions. It also creats
+    a mapping of node and molecule index to the global
+    index in the position matrix as well as a cdk tree
+    of defined positions and a mapping between global
+    indices and defined positions. Furthermore, it stores
+    the interaction matrix between atomtypes and the atom
+    types, where each atomtype corresponds to one of the
+    global indices.
+    """
+
+    def __init__(self,
+                 positions,
+                 nodes_to_idx,
+                 atom_types,
+                 interaction_matrix,
+                 cut_off,
+                 boxsize):
+
+        self.positions = positions
+        self.nodes_to_gndx = nodes_to_idx
+        self.atypes = np.asarray(atom_types, dtype=str)
+        self.interaction_matrix = interaction_matrix
+        self.cut_off = cut_off
+        self.boxsize = boxsize
+
+        self.defined_idxs = [list(np.where(self.positions[:, 0].reshape(-1) != np.inf)[0])]
+        self.position_trees = [scipy.spatial.ckdtree.cKDTree(positions[self.defined_idxs[-1]],
+                                                             boxsize=boxsize)]
+
+    def copy(self):
+        """
+        Return new instance and deep copy of the objects position attribute.
+        """
+        new_obj = NonBondMatrix(self.positions.copy(),
```

Comment:
But that defeats the purpose of this copy method all-together, because it will modify the position_trees in the parent object which is not what you want.